owner-onboarding-server/src/ov_management/mod.rs

```diff
@@ -0,0 +1,256 @@
+        .ownership_voucher_store
+        .destroy_data(&Guid::from_str(&i).unwrap())
+        .await.unwrap();
+    }
+
+    Ok(Box::new(warp::reply()))
+}
+
+
+const ACCEPTED_CONTENT_TYPE:[&str; 2] = ["application/x-pem-file","application/cbor"];
+
+#[derive(Clone)]
+#[derive(Serialize,Deserialize)]
+#[derive(Eq,PartialEq,Hash)]
+#[derive(Error,Debug)]
+enum ErrorCode {
+    #[error("error")]
+    ParseError,
+    #[error("parsed")]
+    InvalidNumberOfVouchers,
+    #[error("unowned_voucher")]
+    UnwonedVoucher,
+    #[error("X-Number-Of-Vouchers")]
+    VoucherCountError,
+    #[error("required")]
+    RequireContentTypeHeader,
+    #[error("required")]
+    NoOwnershipVoucherFound,
+    #[error("unknown")]
+    UnknownDevice,
+    #[error("{0}")]
+    InvalidVoucherSignatureDescription(String),
+    #[error("failed at index")]
+    InvalidVoucherSignature,
+    #[error("{0}")]
+    IncompleteVoucherDescription(String),
+    #[error("parsed correctly")]
+    IncompleteVoucher,
+    #[error("{0}")]
+    PemParseErrorDescription(String),
+    #[error("parsed correctly")]
+    PemParseError,
+}
+
+#[derive(Serialize)]
+struct ErrorMsg {
+    error_code: Option<ErrorCode>,
+    error_details: serde_json::Value
+}
+
+trait Response<T> {
+    fn generate(e: ErrorCode,value: T ) -> core::result::Result<Box<dyn warp::reply::Reply>,warp::Rejection>;
+    fn generate_with_description(e: ErrorCode,value: T,description: &str) -> core::result::Result<Box<dyn warp::reply::Reply>,warp::Rejection>;
+}
+
+impl<T:Serialize> Response<T> for ErrorMsg
+{
+    fn generate(e: ErrorCode,value: T) -> core::result::Result<Box<dyn warp::reply::Reply>,warp::Rejection> 
+    where T: Serialize {
+          let m =     ErrorMsg {
+                error_code:  Some(e.clone()),
+                error_details: json!({e.to_string(): value})
+            };
+        Ok(Box::new(warp::reply::with_status(warp::reply::json(&m), StatusCode::BAD_REQUEST)))
+    }
+    
+    fn generate_with_description(e: ErrorCode,value: T, description: &str) -> core::result::Result<Box<dyn warp::reply::Reply>,warp::Rejection> 
+    where T: Serialize {
+          let m=     ErrorMsg {
+                error_code:  Some(e.clone()),
+                error_details: json!({
+                    "description" : description,
+                    e.to_string(): value
+                })
+            };
+        Ok(Box::new(warp::reply::with_status(warp::reply::json(&m), StatusCode::BAD_REQUEST)))
+    }
+
+}
+
+fn validate_ov<'a> (ov: &'a OwnershipVoucher,
+    owner_key: &'a PKey<Private>,
+) -> Result<bool,ErrorCode> {
+        ov.iter_entries()
+        .map_err(|e| ErrorCode::PemParseErrorDescription(e.to_string()))?
+        .into_iter().next()
+        .unwrap()
+        .map_err(|e| ErrorCode::IncompleteVoucherDescription(e.to_string()))?
+        .public_key()
+        .matches_pkey(owner_key)
+        .map_err(|e| ErrorCode::InvalidVoucherSignatureDescription(e.to_string()))
```

Comment:
Why do you store the `String` version of the `Error` in the `InvalidVoucherSignatureDescription` key?
You could store the full `Error` at this point, and then only turn it to a `String` when actually serializing.

Reply:
tried to modify errorCodes to:
PemParseErrorDescription(fdo_data_formats::Error)

but getting error : trait bound `fdo_data_formats::Error: _::_serde::Serialize` is not satisfied
the trait `_::_serde::Serialize` is not implemented for `fdo_data_formats::Error`

Reply:
dropped this Error as we are reporting custom error message with no much of details